Release 104

benjben opened the pull request:
Required for [release R114](https://github.com/snowplow/snowplow/pull/4076) of Snowplow pipeline.

Contains:
- Updated schemas for Sendgrid
- Schemas for YAUAA enrichment.

alexanderdean commented:
Please assign to me as the last step when everything has been reviewed etc.

chuwy commented on CHANGELOG:
```diff
@@ -1,3 +1,19 @@
+Release 104 (2019-05-14)
+------------------------
+Add nl.basjes/yauaa_context/jsonschema/1-0-0 (closes #934)
```
`closes` is redundant in CHANGELOG.

chuwy commented on schemas/com.sendgrid/bounce/jsonschema/2-0-0:
```diff
@@ -0,0 +1,107 @@
+      "format": "ipv4"
+    },
+    "timestamp": {
+      "description": "the UNIX timestamp of when the message was sent",
```
I think "UNIX timestamp" assumes its an seconds since the epoch, but `date-time` assumes its a ISO8601 string.

dilyand replied:
"UNIX timestamp" comes from the Sendgrid documentation. But as you noted elsewhere, we're schemaing the *output* of the webhook and not the *input*, so the proper format is `date-time`. I have updated the descriptions of the `timestamp` field across all schemas.

chuwy commented on schemas/com.sendgrid/bounce/jsonschema/2-0-0:
```diff
@@ -0,0 +1,107 @@
+    "ip": {
+      "description": "the IP address used to send the email. For open and click events, it is the IP address of the recipient who engaged with the email.",
+      "type": "string",
+      "format": "ipv4"
```
We sure it cannot be ipv6?

dilyand replied:
Yes.

> An Internet Protocol (IP) Address is a unique numerical address that defines an internet location. These addresses have 4 sets of numbers separated by periods. For example, 127.0.0.1 is an IP address.

*Source*: https://sendgrid.com/docs/glossary/ip-address/

chuwy commented on schemas/com.sendgrid/bounce/jsonschema/2-0-0:
```diff
@@ -0,0 +1,107 @@
+    },
+    "email": {
+      "description": "the email address of the recipient",
+      "type": "string"
```
FYI there's `email` format.

chuwy commented on schemas/com.sendgrid/bounce/jsonschema/2-0-0:
```diff
@@ -0,0 +1,107 @@
+      "type": "string"
+    },
+    "newsletter": {
+      "description": "Legacy Marketing Email tool fields",
```
Some descriptions are capitalized, some are not. Can we make it consistent?

chuwy commented on schemas/com.sendgrid/bounce/jsonschema/2-0-0:
```diff
@@ -0,0 +1,107 @@
+      "type": "integer"
+    }
+  },
+  "additionalProperties": true
```
Generally, `additionalProperties: false` is more preferable option as it makes a schema more explicit. It also totally makes sense to make `additionalProperties: true` when you don't have a control over data produced by 3rd party providers. However, in webhooks we're schemaing the _output_ format, not _input_, so if we can assume that webhook can strip away all unkown fields - we certainly can make it `additionalProperties: false`. This is not a request, just food for thought.

chuwy commented on schemas/com.snowplowanalytics.snowplow.enrichments/yauaa_enrichment_config/jsonschema/1-0-0:
```diff
@@ -0,0 +1,38 @@
+            "type": "object",
+            "properties": {
+                "cacheSize": {
+                    "type": "number"
```
Cache size can be non-integer?

benjben replied:
No it can't, I update it

chuwy commented on schemas/nl.basjes/yauaa_context/jsonschema/1-0-0:
```diff
@@ -0,0 +1,192 @@
+        },
+        "deviceBrand": {
+            "description": "Example: Google",
+            "type": "string"
```
I know that there's likely no specification on how long it can be. But in fact we always have an upper bound for strings (at least in Redshift) - it is just implcit 4096. Sometimes we can do better even with guesswork.

benjben replied:
I added `maxLength` for almost all the fields. Hopefully I took enough margin regarding the lengths of the fields, based on their names and some examples.

benjben commented:
Thanks for the review! @dilyand I added you as a reviewer so that you can see the comments related to Sendgrid.

dilyand commented:
@benjben Would you like me to respond to them or just adding me fyi?

benjben commented:
@dilyand If you could respond to them that would be awesome, otherwise no worries I'll do it.

benjben commented:
I updated schemas for YAUAA with `maxLength` and I updated all the descriptions of Sendgrid with a capital letter for consistency. I also used `email` format (and regenerated the SQL code).

For the last comments regarding I let @dilyand comment.

alexanderdean commented:
Please assign to me as the last step when all reviewers have green ticked.

alexanderdean commented:
All done now
